src/app/setup-wizard/data-collection/on-site-assessment/neb-forms-accordion/neb-setup-form/performance-metrics-modal/performance-metrics-modal.component.ts

```diff
@@ -0,0 +1,132 @@
+
+  performanceMetricOptions: Array<KeyPerformanceMetricOption> = [];
+
+  kpmSearchStr: string = '';
+  orderByDir: 'asc' | 'desc' = 'asc';
+  filterAssociatedMetrics: boolean = false;
+  keyPerformanceIndicators: Array<IdbKeyPerformanceIndicator>;
+
+  constructor(private keyPerformanceIndicatorIdbService: KeyPerformanceIndicatorsIdbService,
+    private keyPerformanceMetricImpactIdbService: KeyPerformanceMetricImpactsIdbService
+  ) {
+
+  }
+
+  openMetricModal() {
+    this.keyPerformanceIndicators = this.keyPerformanceIndicatorIdbService.keyPerformanceIndicators.getValue();
+    this.displayMetricsModal = true;
+    this.setMetricOptions();
+  }
+
+  closeAddMetricModal() {
+    this.displayMetricsModal = false;
+    this.performanceMetricToAdd = undefined;
+  }
+
+  async confirmAddMetric() {
+    //make sure metric is tracked in KPI
+    let addedMetric: KeyPerformanceMetric;
+    let keyPerformanceIndicator: IdbKeyPerformanceIndicator = this.keyPerformanceIndicatorIdbService.getKpiFromKpm(this.nonEnergyBenefit.companyId, this.performanceMetricToAdd.kpiValue);
+    if (keyPerformanceIndicator) {
+      addedMetric = keyPerformanceIndicator.performanceMetrics.find(_metric => {
+        return (_metric.value == this.performanceMetricToAdd.value);
+      });
+      await this.keyPerformanceIndicatorIdbService.asyncUpdate(keyPerformanceIndicator);
+    } else {
+      //add untracked KPI if doesn't exist
+      let kpiOption: KeyPerformanceIndicatorOption = KeyPerformanceIndicatorOptions.find(option => {
+        return option.optionValue == this.performanceMetricToAdd.kpiValue
+      });
+      keyPerformanceIndicator = getNewKeyPerformanceIndicator(this.nonEnergyBenefit.userId, this.nonEnergyBenefit.companyId, kpiOption, false);
+      addedMetric = keyPerformanceIndicator.performanceMetrics.find(_metric => {
+        return (_metric.value == this.performanceMetricToAdd.value);
+      });
+      await firstValueFrom(this.keyPerformanceIndicatorIdbService.addWithObservable(keyPerformanceIndicator));
+      await this.keyPerformanceIndicatorIdbService.setKeyPerformanceIndicators();
+    }
+
+    let newKeyPerformanceMetricImpact: IdbKeyPerformanceMetricImpact = getNewIdbKeyPerformanceMetricImpact(this.nonEnergyBenefit.userId, this.nonEnergyBenefit.companyId, this.nonEnergyBenefit.facilityId, this.nonEnergyBenefit.energyOpportunityId, this.nonEnergyBenefit.guid, addedMetric.value, this.nonEnergyBenefit.assessmentId, keyPerformanceIndicator.guid, addedMetric.guid);
+    await firstValueFrom(this.keyPerformanceMetricImpactIdbService.addWithObservable(newKeyPerformanceMetricImpact));
+    await this.keyPerformanceMetricImpactIdbService.setKeyPerformanceMetricImpacts();
+    this.closeAddMetricModal();
+  }
+
+  addMetric(performanceMetric: KeyPerformanceMetricOption | KeyPerformanceMetric) {
+    this.performanceMetricToAdd = performanceMetric;
+  }
+
+  setMetricOptions() {
```

Comment:
It seems the same KPM can be added multiple times to the same NEB. Not sure if this is the behavior we want.

Reply:
whoops. Fixed this.